packages/loopring_v3/contracts/amm/AmplifiedAmmController.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+
+import "./IAmmController.sol";
+import "../amm/LoopringAmmPool.sol";
+import "../lib/Claimable.sol";
+import "../lib/MathUint.sol";
+import "../lib/MathUint96.sol";
+import "../thirdparty/SafeCast.sol";
+
+
+/// @author Brecht Devos - <brecht@loopring.org>
+contract AmplifiedAmmController is IAmmController, Claimable
+{
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+
+    uint public constant AMPLIFICATION_FACTOR_BASE = (10 ** 18);
+
+    mapping(address => uint) public amplificationFactors;
+
+    function getInitialVirtualBalances(
+        uint96[] memory joinAmounts
+        )
+        external
+        view
+        override
+        returns (uint96[] memory)
+    {
+        uint amplificationFactor = getAmplificationFactor(msg.sender);
+        uint96[] memory vTokenBalancesL2 = new uint96[](joinAmounts.length);
+        for (uint i = 0; i < joinAmounts.length; i++) {
+            vTokenBalancesL2[i] = (uint(joinAmounts[i]).mul(amplificationFactor) / AMPLIFICATION_FACTOR_BASE).toUint96();
+        }
+        return vTokenBalancesL2;
+    }
+
+    function getVirtualBalances(
+        uint96[] memory tokenBalancesL2,
+        uint96[] memory /*vTokenBalancesL2*/
+        )
+        external
+        view
+        override
+        returns (uint96[] memory)
+    {
+        // Only allow updating the virtual balances if the AF = 1
+        require(getAmplificationFactor(msg.sender) == AMPLIFICATION_FACTOR_BASE, "INVALID_OPERATION");
+
+        // Just set the virtual balances to the actual balances
+        uint96[] memory vTokenBalancesL2 = new uint96[](tokenBalancesL2.length);
+        for (uint i = 0; i < tokenBalancesL2.length; i++) {
+            vTokenBalancesL2[i] = tokenBalancesL2[i];
```

Comment:
The current implement seems to **reset** the virtual balances to the same value as the actual balances. Why not allow the owner to set them to arbitrary values (while keep the price)

Reply:
If AF == 1 then the virtual balances are the actual balances, so this can be done without changing the behavior of the pool and this is what will need to be done to update our current pools to the new virtual balance system, which is the main reason this function is implemented like it is. If AF != 1 then the pool was already created in the new system and no updates are necessary to the virtual balances after the pool has been created.

Allowing the virtual balances to be updated using different pools could be valid, but it won't be a normal amplified pool so I did not make it possible. Allowing the curve to change while funds are already in the pool will impact things like IL even when price remains the same so it's more of an actively managed pool where people would need to trust whatever controls what happens with the pool.